### ISSUE TITLE
Some potentially useful operations, open for discussion

### ISSUE THREAD

**Issue by jwa7** [lab-cosmo/equistore#145]
These suggestions are motivated by the desire to have convenience functions that do the following jobs and eliminate developer time hard coding them. This issue aims to be a starting point for discussion; the function names, exact functionalities, whether they are in place or not, or indeed whether they are actually useful (!) etc., are up for debate. I'm sure there are also some technicalities with memory allocation and immutability.

### 1. `insert` or `update`

A function to update, perhaps in a similar style to update with Python dictionaries, a TensorMap with new key/block pairs, provided the new keys are not currently in the existing TensorMap

```py
from equistore.operations import insert_blocks

tensor = io.load(...)

# Define new blocks to be inserted
blocks_to_insert = [TensorBlock(...), TensorBlock(...)]

# Insert the blocks
new_tensor = insert_blocks(tensor, keys=Labels(names=["key_1", "key_2"], values=np.array([[5, 8],]), blocks=blocks_to_insert)
```

### 2. `drop_blocks`

Suppose I want to remove a block by its key name. Assuming this returns a new tensor:

```py
from equistore.operations import drop_blocks

tensor = equistore.io.load(...)

# tensor with 12 blocks
tensor
>>>
TensorMap with 12 blocks
keys: ['spherical_harmonics_l' 'species_center']
                  0                   1
                  1                   1
                  2                   1
               ...
                  3                   8
                  4                   8
                  5                   8

new_tensor = drop_blocks(
    tensor, 
    keys=Labels(
        names=["spherical_harmonics_l", "species_center"],
        values=np.array([[2, 1], [3, 8]]),
    )
)

# Now has 10 blocks
new_tensor
>>>
TensorMap with 10 blocks
keys: ['spherical_harmonics_l' 'species_center']
                  0                   1
                  1                   1
               ...
                  4                   8
                  5                   8
```

The `drop_blocks` function presented here could be viewed like a kind of `slice` but for key/block pairs. Should then it work like `slice`, specifying the keys that should be **kept**, as opposed to the ones that should be **dropped**?

### 3. `merge` or `unify`  or `join_along_keys`

Similar to `join()` but doesn't change the blocks themselves or join along a block axis. Just takes 2 TensorMaps and returns a new larger TensorMap whose keys are just the union of the keys of the original TensorMaps. Assumes that the intersection in key names between the 2 (or in principle more) TensorMaps is empty.

```py
...
# First tensor with a certain set of keys
tensor_1
>>>
TensorMap with 7 blocks
keys: ['spherical_harmonics_l' 'species_center']
                  0                   1
               ...
                  5                   8

# second tensor with a different set of keys
tensor_2
>>>
TensorMap with 10 blocks
keys: ['spherical_harmonics_l' 'species_center']
                  1                   1
               ...
                  4                   8

# merged tensor with keys and blocks from both
merged_tensor = merge(tensors=[tensor_1, tensor_2])
merged_tensor
>>>
TensorMap with 17 blocks
keys: ['spherical_harmonics_l' 'species_center']
                  0                   1
                  1                   1
               ...
                  4                   8
                  5                   8

```

#### a) keys

Suppose that I have dropped some blocks in my TensorMap such that some of the values of the key names are the same and therefore redundant. 

As an example: generating a lambda-SOAP descriptor for both even (+1) and odd (-1) inversion parities. If all blocks with uneven parity are dropped we are left with blocks that have a redundant ``"inversion_sigma"=+1`` key name. We therefore might want to drop the key name "inversion_sigma". This function would assume that the value for the chosen name(s) is equivalent for all keys.

```py
from equistore.operations import drop_metadata_names

tensor = io.load(...)

# tensor with 10 blocks but a redundant key name
tensor
>>>
TensorMap with 10 blocks
keys: ['inversion_sigma', 'spherical_harmonics_l' 'species_center']
                  1                   0                   1
                  1                   1                   1
               ...
                  1                   4                   8
                  1                   5                   8

# remove redundant key name
cleaner_tensor = drop_metadata_names(tensor, axis="keys", names=["inversion_sigma"])
cleaner_tensor
>>>
TensorMap with 10 blocks
keys: ['spherical_harmonics_l' 'species_center']
                  0                   1
                  1                   1
               ...
                  4                   8
                  5                   8
```

#### b) samples/properties

Now suppose I have a TensorMap where I have used `slice` to include, for instance only 1 structure. The following example is one I have encountered when using external codes (i.e. [Q-Stack/equio.py](https://github.com/lcmd-epfl/Q-stack/blob/master/qstack/equio.py)) and have needed to drop the structure label from my samples axis. Again this assumes that the values of the named metadata along the given axis is equivalent for all blocks.

```py
from equistore.operations import drop_metadata_names
...
# only structure index zero exists in this (and all) blocks
tensor.block(0).samples
>>>
Labels([( 0, 1), ( 0, 2), ( 0, 3), ( 0, 4)], dtype=[('structure', '<i4'), ('center', '<i4')])

new_tensor = drop_metadata_names(tensor, axis="samples", names=["structure"])

# remove the redundant samples name "structure"
new_tensor.block(0).samples
>>>
Labels([( 1,), ( 2,), (3,), (4,)], dtype=[('center', '<i4')])
```
and the equivalent along properties.

### 5. `insert_metadata_names`

As the inverse of `drop_metadata_names`, suppose we want to introduce either keys/samples/properties names of fixed values into a TensorMap.

#### a) keys

Suppose I want to use the `merge` function from above but with 2 TensorMaps with identical metadata. It could be useful to introduce a new key name that distinguishes them, before merging:

```py
...
# first and second tensors have identical keys 
tensor_1
>>>
TensorMap with 10 blocks
keys: ['spherical_harmonics_l' 'species_center']
                  0                   1
                  1                   1
               ...
                  4                   8
                  5                   8

tensor_2
>>>
TensorMap with 10 blocks
keys: ['spherical_harmonics_l' 'species_center']
                  0                   1
                  1                   1
               ...
                  4                   8
                  5                   8

# Insert a new key name to distinguish the 2 TensorMaps
new_tensor_1 = insert_key_names(tensor_1, names=["inversion_sigma"], values=[+1], prepend=True)
new_tensor_2 = insert_key_names(tensor_2, names=["inversion_sigma"], values=[-1], prepend=True)

new_tensor_1
>>>
TensorMap with 10 blocks
keys: ['inversion_sigma', 'spherical_harmonics_l' 'species_center']
                  1                   0                   1
                  1                   1                   1
               ...
                  1                   4                   8
                  1                   5                   8

new_tensor_2
>>>
TensorMap with 10 blocks
keys: ['inversion_sigma', 'spherical_harmonics_l' 'species_center']
                  -1                   0                   1
                  -1                   1                   1
               ...
                  -1                   4                   8
                  -1                   5                   8

# Now merge safely
merged_tensor = merge([new_tensor_1, new_tensor_2])
```

Where `prepend=False` would instead append the new key name to the list of key names.

#### b) samples/properties

When using the `join` function as currently implemented, the function attempts to resolve metadata conflicts should there be any. If we have 2 TensorMaps with exactly the same metadata, we could first introduce another samples name (the naming of which we have control) to the blocks in each respective TensorMap, with different values, and then join safely.

```py
from equistore.operations import insert_metadata_names
...
tensor_1.block(0).samples
>>>
Labels([( 1,), ( 2,), (3,), (4,)], dtype=[('center', '<i4')])

tensor_2.block(0).samples
>>>
Labels([( 1,), ( 2,), (3,), (4,)], dtype=[('center', '<i4')])

new_tensor_1 = insert_metadata_names(tensor, axis="samples", names=["structure"], values=[0,], prepend=True)
new_tensor_1.block(0).samples
>>>
Labels([( 0, 1), ( 0, 2), ( 0, 3), ( 0, 4)], dtype=[('structure', '<i4'), ('center', '<i4')])

new_tensor_2 = insert_metadata_names(tensor, axis="samples", names=["structure"], values=[1,], prepend=True)
new_tensor_2.block(0).samples
>>>
Labels([( 1, 1), ( 1, 2), ( 1, 3), ( 1, 4)], dtype=[('structure', '<i4'), ('center', '<i4')])

joined_tensor = join([new_tensor_1, new_tensor_2], axis="samples")
```

Where `prepend=False` would instead append the new sample name to the list of sample names.

### 6. inverse operations of ``keys_to_x`` etc?

Would it be useful to have the inverse of `keys_to_properties` and `keys_to_samples`, and `components_to_properties`? (Not come across a use for these personally, but think I've heard it mentioned before by people)

**Comment by jwa7** [comment]
### 7. `absolute`

Analogous to `numpy.absolute()` or `torch.abs()`, takes a TensorMap and returns a TensorMap of the same metadata, but where the blocks contain the absolute values of the original blocks

**Comment by jwa7** [comment]
I have some code that does some of these jobs. Posting it here incase anyone needs it before a PR is created. Note: prototype stage, and untested.

**drop_blocks**

```py
def drop_blocks(tensor: TensorMap, keys: Labels) -> TensorMap:
    """
    Drop specified key/block pairs from a TensorMap.

    :param tensor: the TensorMap to drop the key-block pair from.
    :param keys: the key Labels of the blocks to drop

    :return: the input TensorMap with the specified key/block pairs dropped.
    """
    if not np.all(tensor.keys.names == keys.names):
        raise ValueError(
            "The input tensor's keys must have the same names as the specified"
            + f" keys to drop. Should be {tensor.keys.names} but got {keys.names}"
        )
    new_keys = tensor.keys[tensor.keys != keys]
    return TensorMap(keys=new_keys, blocks=[tensor[key].copy() for key in new_keys])
```

**drop_key_name**

```py
def drop_key_name(tensor: TensorMap, key_name: str) -> TensorMap:
    """
    Takes a TensorMap and drops the key_name from the keys. Every key must have
    the same value for the key_name, otherwise a ValueError is raised.
    """
    # Check that the key_name is present and unique
    if not len(np.unique(tensor.keys[key_name])) == 1:
        raise ValueError(
            f"key_name {key_name} is not unique in the keys."
            + " Can only drop a key_name where the value is the"
            + " same for all keys."
        )

    # Define the idx of the key_name to drop
    drop_idx = tensor.keys.names.index(key_name)

    # Build the new keys
    new_keys = Labels(
        names=tensor.keys.names[:drop_idx] + tensor.keys.names[drop_idx + 1 :],
        values=np.array(
            [k.tolist()[:drop_idx] + k.tolist()[drop_idx + 1 :] for k in tensor.keys]
        ),
    )
    return TensorMap(keys=new_keys, blocks=[b.copy() for b in tensor.blocks()])
```


**drop_metadata_name**

```py
def drop_metadata_name(tensor: TensorMap, axis: str, name: str) -> TensorMap:
    """
    Takes a TensorMap and drops the `name` from either the "samples" or
    "properties" labels of every block. Every block must have the same value for
    the `name`, otherwise a ValueError is raised.
    """
    if axis not in ["samples", "properties"]:
        raise ValueError(f"axis must be 'samples' or 'properties', not {axis}")
    # Check that the name is present and unique
    for block in tensor.blocks():
        if axis == "samples":
            uniq = np.unique(block.samples[name])
        else:
            uniq = np.unique(block.properties[name])
        if not len(uniq) == 1:
            raise ValueError(
                f"name {name} is not unique in the {axis}."
                + " Can only drop a `name` where the value is the"
                + f" same for all {axis}."
            )
    # Identify the idx of the name to drop
    if axis == "samples":
        drop_idx = tensor.blocks()[0].samples.names.index(name)
    elif axis == "properties":
        drop_idx = tensor.blocks()[0].properties.names.index(name)
    # Construct new blocks with the dropped name
    new_blocks = []
    for key in tensor.keys:
        new_samples = tensor[key].samples
        new_properties = tensor[key].properties
        if axis == "samples":
            new_samples = Labels(
                names=tensor[key].samples.names[:drop_idx]
                + tensor[key].samples.names[drop_idx + 1 :],
                values=np.array(
                    [
                        s.tolist()[:drop_idx] + s.tolist()[drop_idx + 1 :]
                        for s in tensor[key].samples
                    ]
                ),
            )
        else:
            new_properties = Labels(
                names=tensor[key].properties.names[:drop_idx]
                + tensor[key].properties.names[drop_idx + 1 :],
                values=np.array(
                    [
                        p.tolist()[:drop_idx] + p.tolist()[drop_idx + 1 :]
                        for p in tensor[key].properties
                    ]
                ),
            )
        new_blocks.append(
            TensorBlock(
                new_samples, tensor[key].components, new_properties, tensor[key].values
            )
        )
    return TensorMap(
        keys=tensor.keys,
        blocks=new_blocks,
    )
```

**Comment by Luthaf** [comment]
> ### 6. inverse operations of ``keys_to_x`` etc?
>
> Would it be useful to have the inverse of `keys_to_properties` and `keys_to_samples`, and `components_to_properties`? (Not come across a use for these personally, but think I've heard it mentioned before by people)

See #186

**Comment by Luthaf** [comment]
The metadata manipulation was merged in #299. The only remaining idea in this issue which is not implemented is `add_block`, which is now tracked in #312
